Posts/2023/06/Measuring-Download-Time.md

```diff
@@ -0,0 +1,544 @@
+
+### `Start-BitsTransfer`
+
+BITS, or Background Intelligent Transfer Service is a Windows service that manages content transfer
+using HTTP or SMB. It was designed to manage the many aspects of content transfer, including cost,
+speed, priority, etc.
+For us, it also serves as an easy way of downloading files. Here is how you download a file from a
+web server using BITS:
+
+```powershell
+Start-BitsTransfer -Source 'https://www.myawesomeserver.com/Files/BitsDefinition.txt' `
+-Destination 'C:\BitsDefinition.txt'
+```
+
+Another great advantage of BITS is that it shows progress, which can be useful while
+downloading big files. In our case however, we want to know how long does it take to
+download a file. For this we will use a handy object of type `System.Diagnostics.Stopwatch`.
+
+```powershell
+$stopwatch = New-Object -TypeName 'System.Diagnostics.Stopwatch'
+
+$stopwatch.Start()
+Start-BitsTransfer -Source 'https://www.myawesomeserver.com/Files/BitsDefinition.txt' `
+-Destination 'C:\BitsDefinition.txt'
+$stopwatch.Stop()
+
+Write-Output $stopwatch.Elapsed
+```
+
+```powershell-console
+Days              : 0
+Hours             : 0
+Minutes           : 0
+Seconds           : 0
+Milliseconds      : 816
+Ticks             : 8165482
+TotalDays         : 9.45078935185185E-06
+TotalHours        : 0.000226818944444444
+TotalMinutes      : 0.0136091366666667
+TotalSeconds      : 0.8165482
+TotalMilliseconds : 816.5482
+```
+
+Awesome, we now have a baseline to build our script upon. First thing we will change is the file,
+since we are more interested on the speed, we can use temporary files to download, that also
+gives us the opportunity of cleaning up at the end. For this we will use a static method from
+`System.IO.Path` called `GetTempFileName`. Other thing we must think is on running the test a number
+of times, and calculating the average, this way we have more reliable results.
+
+```powershell
+# Changing the progress preference to hide the progress bar.
+$ProgressPreference = 'SilentlyContinue'
+$payloadUrl = 'https://www.myawesomeserver.com/Files/BitsDefinition.txt'
+$destination = 'C:\BitsDefinition.txt'
+$stopwatch = New-Object -TypeName 'System.Diagnostics.Stopwatch'
+$elapsedTime = [timespan]::Zero
+$iterationNumber = 3
+
+# Here we are using a foreach loop with a range,
+# but this can also be accomplished with a for loop.
+foreach ($iteration in 1..$iterationNumber) {
+    $tempFilePath = [System.IO.Path]::GetTempFileName()
+    
+    $stopwatch.Restart()
+    Start-BitsTransfer -Source $payloadUrl -Destination $destination
+    $stopwatch.Stop()
+    
+    Remove-Item -Path $tempFilePath
+    $elapsedTime = $elapsedTime.Add($stopwatch.Elapsed)
+}
+
+# Timespan.Divide is not available on .NET Framework.
+if ($Host.Version -ge [version]'6.0') { $average = $elapsedTime.Divide($IterationNumber) }
+else { $average = [timespan]::new($elapsedTime.Ticks / $IterationNumber) }
+
+return $average
```

Comment:
Fixed the following:
- hostname
- You created the tempfile name but never used it
- You should test the version of PowerShell not the version of the Host because you don't know what host it will be running in.

```suggestion
# Changing the progress preference to hide the progress bar.
$ProgressPreference = 'SilentlyContinue'
$payloadUrl = 'https://www.contoso.com/Files/BitsDefinition.txt'
$stopwatch = New-Object -TypeName 'System.Diagnostics.Stopwatch'
$elapsedTime = [timespan]::Zero
$iterationNumber = 3

# Here we are using a foreach loop with a range,
# but this can also be accomplished with a for loop.
foreach ($iteration in 1..$iterationNumber) {
    $tempFilePath = [System.IO.Path]::GetTempFileName()

    $stopwatch.Restart()
    Start-BitsTransfer -Source $payloadUrl -Destination $tempFilePath
    $stopwatch.Stop()

    Remove-Item -Path $tempFilePath
    $elapsedTime = $elapsedTime.Add($stopwatch.Elapsed)
}

# Timespan.Divide is not available on .NET Framework.
if ($PSVersionTable.PSVersion -ge [version]'6.0') {
    $average = $elapsedTime.Divide($IterationNumber)
} else { $
    average = [timespan]::new($elapsedTime.Ticks / $IterationNumber)
}

return $average
```